### PR TITLE
perf(replay): reduce DOM accessor overhead and add benchmarks

### PR DESCRIPTION
**Pull request by marandaneto**
## Problem

Large session replay snapshots and structural mutations still block the main thread on node-dense pages. Profiling the built SDK found that rrweb repeatedly constructs composite string keys when looking up its cached native DOM getters.

This is a small first improvement for #4217, not a complete fix. DOM capture and parts of payload encoding remain synchronous.

## Changes

- Group the native getter cache by prototype so node accesses reuse property keys instead of constructing strings. Keep native getter selection, receiver binding and instance fallback unchanged. The cache still holds getters, not DOM or masking state.
- Add regression tests for repeated key conversion, changing DOM values and receivers, patched getters, shadow-root access and inherited property names.
- Add the opt-in `pnpm benchmark:replay` command. It exercises the built SDK through mocked replay transport with compression on and off, checks intermediate replay state and privacy sentinels, and reports long tasks, frame gaps, CPU, heap deltas, drops and extra full snapshots.
- Add dense-DOM and CSSOM/adopted-stylesheet workloads with startup, full snapshot, rebuild, move and removal phases. Document configuration, measurement boundaries and remaining coverage gaps.

### Performance results

Three unprofiled repetitions per arm on a roughly 50k-node fixture in Playwright Chromium 136.0.7103.25, with no CPU throttling. Both builds use the same generation-aware benchmark. These are descriptive local samples, not statistically significant estimates or customer-level guarantees.

Median longest observed main-thread task:

| Workload | Compression enabled, before / after | Compression disabled, before / after |
| --- | --- | --- |
| Start recording | 129 / 121 ms | 286 / 264 ms |
| Full snapshot | 105 / 97 ms | 119 / 109 ms |
| Rebuild subtree | 275 / 250 ms | 290 / 291 ms |
| Move subtree | 233 / 220 ms | 257 / 234 ms |

Uncompressed rebuild performance was essentially unchanged. No oversized mutation drops were observed in this comparison, and replay-state and privacy assertions passed. The recorder artifact increased by 110 bytes before compression and 27 bytes with gzip.

The benchmark records artifact hashes and environment metadata. It deliberately separates profiled runs from timing comparisons. A zero longest-task result means no observed task reached the Long Tasks API's 50 ms threshold, not zero blocking.

### Validation

- Browser SDK and dependency build passed.
- rrweb-utils typecheck and targeted lint/format checks passed.
- Recording and accessor suites: 316 tests passed, 2 skipped across 23 files.
- Browser masking suite: 9 tests passed across Chromium, Firefox and WebKit.
- Table and CSS benchmark smoke tests passed with both compression settings.
- A temporary fault injection that discarded the rebuild mutation caused the rebuild checkpoint to fail as expected.
- Isolated Pi autoreview passed on `e9058295e8b060a9c847bdb4f0872fa144e041be`, with no actionable findings.

No recorder configuration, event ordering, masking policy, sampling or session lifecycle behavior changes. No recording-volume change is expected. Encoding-tail and asynchronous snapshot work will be evaluated separately.

### Libraries affected

- [ ] All of them
- [x] posthog-js (web)
- [ ] posthog-js-lite (web lite)
- [ ] posthog-node
- [ ] posthog-react-native
- [ ] @posthog/react-native-plugin
- [ ] @posthog/react
- [ ] @posthog/ai
- [ ] @posthog/convex
- [ ] @posthog/next
- [ ] @posthog/nextjs-config
- [ ] @posthog/nuxt
- [ ] @posthog/openfeature-node-provider
- [ ] @posthog/openfeature-web-provider
- [ ] @posthog/rollup-plugin
- [ ] @posthog/webpack-plugin
- [ ] @posthog/types
- [ ] @posthog/browser-common

## Checklist

- [x] Tests for new code
- [x] Accounted for the impact of any changes across different platforms
- [x] Accounted for backwards compatibility of any changes (no breaking changes!)
- [x] Took care not to unnecessarily increase the bundle size

### If releasing new changes

- [x] Added a `posthog-js` patch changeset

## 🤖 Agent context

**Autonomy:** Human-driven (agent-assisted)

Implemented with Pi using repository inspection, Chromium profiling, focused tests and an independent review. The work keeps synchronous recording semantics and avoids importing the time-sliced snapshot implementation from the earlier closed PRs.

The pre-commit code formatter attempted to reformat entire rrweb files, although the root formatting command excludes rrweb. Only that hook was skipped to retain the existing rrweb style. Other hooks and targeted formatting checks passed. No changes were made to hook configuration.

### CODE REVIEW

**Comment by marandaneto** [comment]
not a considerable improvement but its still worth since its a small change

**Review by fasyy612 — APPROVED** [review]

File: packages/rrweb/utils/src/index.ts [hunk 1]
```diff
@@ -176,7 +183,7 @@
 
   if (!untaintedAccessor) return instance[accessor];
 
-  untaintedAccessorCache[cacheKey] = untaintedAccessor;
+  cache[accessor as string] = untaintedAccessor;
```
[the remark this answers is not in the document]
Comment (marandaneto) [reply packages/rrweb/utils/src/index.ts:187]: Hardened in 5bdca092a20a4a45b8486fd79d0ab86661ae0c1d: the outer prototype-to-bucket table now also has a null prototype. Three red/green tests reproduce the inherited-key path: __proto__ previously selected Object.prototype and invoked its setter, while constructor/toString selected inherited functions. The engine rejected the prototype mutation in the reproduction; this is not a claim of a successful pollution exploit. All three now reject invalid bucket names before touching inherited objects, while special accessor names within a valid bucket retain their normal fallback. Validation: 319 recording/accessor tests (2 existing skips), 9 cross-browser masking cases, build/lint/ES checks and committed-branch autoreview passed. CodeQL is being rerun on this head.

File: packages/browser/scripts/benchmark-replay.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
+# Replay main-thread benchmark (#4217)
+
+An **opt-in** benchmark of the built browser SDK, from page work through actual
+mocked replay requests. Uses the installed Playwright Chromium (no external sites
+or new dependencies). Timing numbers are evidence, not machine-independent CI
```
[the remark this answers is not in the document]
Comment (marandaneto) [reply packages/browser/scripts/benchmark-replay.md:5]: I checked the repository instructions and the cited PostHog/posthog#32520. That PR reports before/after averages and standard deviations; it does not establish mandatory timing thresholds for this opt-in diagnostic tool. This harness deliberately reports hardware- and workload-dependent measurements, while failing on recorded-data fidelity violations. The optimization also has a deterministic regression test: 200 warmed DOM accesses must perform zero explicit String conversions, which fails with the original composite-key implementation. I do not think we should invent a latency threshold from a few local desktop samples or treat a descriptive median as statistical significance. A controlled-runner, baseline-relative performance gate could be a separate follow-up once its variance and acceptance criteria are established. Leaving this thread open for agreement rather than resolving a disagreement unilaterally.

**Comment by marandaneto** [comment]
Review follow-up validation for 5bdca092a (relative to the previously reviewed e9058295e): three alternating pairs at ~50k table nodes, identical unprofiled harness, compression on/off, no concurrent builds/tests. All 12 arms passed decoded-wire and replay fidelity checks.

Longest-task medians in ms, before → after:

| Compression | Start | Snapshot | Rebuild | Move |
| --- | --- | --- | --- | --- |
| on | 120 → 124 | 92 → 92 | 269 → 267 | 222 → 231 |
| off | 265 → 263 | 117 → 116 | 277 → 282 | 250 → 238 |

These are mixed, overlapping local samples, not a new performance win or proof of equivalence. The zero-explicit-String-conversion work-count regression still passes. Cache hardening adds 186 raw / 55 gzip bytes relative to e9058295e.

Separately, the existing wireMs metric is invalid on the mocked routes in these runs: Playwright request.timing().startTime is zero, so subtracting epochStart yields a negative value. It was not used in the comparisons above. The harness is unchanged in this commit; that timing-metric issue needs a separate follow-up.

File: packages/rrweb/utils/src/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (marandaneto) [reply packages/rrweb/utils/src/index.ts:187]: The CodeQL rerun passed: https://github.com/PostHog/posthog-js/actions/runs/34113747820 . Alert 5391 is now marked fixed for refs/pull/4801/merge.
